### PR TITLE
[Wallet]: Historic Gold prices chart

### LINKED ISSUE #1341: Users SBAT historic gold prices in gold tab
### Expected Behavior

Gold Tab

[Figma](https://www.figma.com/file/Iqu4mAgePHVgKhvKTrGrUP/Gold-Approaches?node-id=10%3A0)

Scope:
* add graphs with historical price + portfolio value
* [out of scope] color changes, design system update coming soon.

To Be decided in kickoff:
- What price granularity is available in charts?
- Which, if any, timeframes to enable/display on graphic initially?

### Current Behavior

What is the current behavior?

### PR DESCRIPTION
**Pull request by i1skn**
### Description

This PR adds historic Gold prices as a chart.

- [x] Add a chart to Gold tab
- [x] Store gold prices from firebase in redux
- [x] Sync only new gold prices when app is restarted
- [x] Show aggregated prices on the chart depend on a chosen range (currently hardcoded).
- [x] Show current value and change in percentage.
- [x] Last point on the chart should come from API.
- [x] Grey line with the current price on the chart
- [x] Lowest/highest values on the chart
- [x] Historic data <- Firebase; Current Rate <- API. Loaders should be synchronized.
- [x] Tests

### Caveats
* Max/Min price are not in the rounded box with shadows - it's very tricky to implement, so, I hope it's not a deal braker for the first iteration on this one.
* Timeframes were taken out of the scope and so far the only one - 30 days is hardcoded. Implementation is flexible for adding them later.

### Tested

Run on Android and iOS device

### Other changes

Little refactoring of how we convert local currencies and some small fixes on the way

### Related issues

- Fixes #1341

### Backwards compatibility

Yes

### Screenshots
154RqDYDN3fGQFz3fZtKthjmRVfS4hKv

### CODE REVIEW

**Comment by cmcewen** [comment]
How did you pick which charting library to use? I have used https://formidable.com/open-source/victory/ before - just curious if you evaluated any other options

**Comment by i1skn** [comment]
@cmcewen I've take a look on some options (but very shallowly tbh) out there and pick this one for a couple of reasons:
* It's fairly simple to understand how it works from the inside (it is pretty small)
* It's easy to configure it the way we need, cause our design is a bit non-standard

I do not have a strong preference though. We can use other options if somebody can recommend any. I just do not have prior experience working with charts in RN.

**Comment by i1skn** [comment]
Also, I'm happy to move implementation to victory charts, as @cmcewen and @jeanregisser both have satisfying experience with it. But, I would prefer to do it in a separate PR, so we can start to iterate on this feature sooner.

File: packages/mobile/locales/en-US/exchangeFlow9.json [hunk 1]
```diff
@@ -1,5 +1,7 @@
 {
   "learnMore": "Learn More",
+  "celoGold": "Celo Gold",
```
Comment (jmrossy) [inline packages/mobile/locales/en-US/exchangeFlow9.json]: Use the existing "gold" translation in global.json

File: packages/mobile/locales/es-419/exchangeFlow9.json [hunk 2]
```diff
@@ -1,5 +1,7 @@
 {
   "learnMore": "Saber más",
+  "celoGold": "~~Celo Gold",
+  "equalToAmount": "~~Equal to {{amount}}",
```
Comment (jmrossy) [inline packages/mobile/locales/es-419/exchangeFlow9.json]: For most small strings, feel free to use google translate for now

File: packages/mobile/locales/es-419/global.json [hunk 3]
```diff
@@ -107,5 +107,8 @@
   "processing": "Procesando",
   "moreWithCount": "~~+{{count }} more",
   "unknown": "~~Unknown",
-  "emptyList": "~~Empty list"
+  "emptyList": "~~Empty list",
```
Comment (jmrossy) [inline packages/mobile/locales/es-419/global.json]: ditto

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+} from 'react-native'
+import { LineChart } from 'react-native-chart-kit'
+import { useDispatch } from 'react-redux'
+import { syncCeloGoldExchangeRateHistory } from 'src/exchange/actions'
+import { Namespaces, withTranslation } from 'src/i18n'
+import { LocalCurrencyCode } from 'src/localCurrency/consts'
+import { useDollarsToLocalAmount, useLocalCurrencyCode } from 'src/localCurrency/hooks'
+import useSelector from 'src/redux/useSelector'
+import { getLocalCurrencyDisplayValue } from 'src/utils/formatting'
+
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: Nit, spacing here. Can you configure vim to organize imports in the same way as vscode?

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+import { Circle, G, Line, Text as SvgText } from 'react-native-svg'
+import { useGoldToDollarAmount } from 'src/exchange/hooks'
+import { exchangeHistorySelector } from 'src/exchange/reducer'
+
+const CHART_POINTS_NUMBER = 60
+const CHART_WIDTH = Dimensions.get('window').width - variables.contentPadding * 2
+const CHART_HEIGHT = 180
+
+const styles = StyleSheet.create({
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: Nit we usually put styles under the JSX but it's no big deal

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+import { useDollarsToLocalAmount, useLocalCurrencyCode } from 'src/localCurrency/hooks'
+import useSelector from 'src/redux/useSelector'
+import { getLocalCurrencyDisplayValue } from 'src/utils/formatting'
+
+import { Circle, G, Line, Text as SvgText } from 'react-native-svg'
+import { useGoldToDollarAmount } from 'src/exchange/hooks'
+import { exchangeHistorySelector } from 'src/exchange/reducer'
+
+const CHART_POINTS_NUMBER = 60
+const CHART_WIDTH = Dimensions.get('window').width - variables.contentPadding * 2
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: any reason to not use variables.width instead of Dimensions here?

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    [localCurrencyCode]
+  )
+  const goldToDollars = useCallback(useGoldToDollarAmount, [])
+  const dollarsToLocal = useCallback(useDollarsToLocalAmount, [])
+
+  const [range] = useState(30 * 24 * 60 * 60 * 1000) // 30 days
+  useEffect(() => {
+    const interval = setInterval(() => {
+      dispatch(syncCeloGoldExchangeRateHistory())
+    }, 600000) // Sync history every 10 minutes
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: long intervals are discouraged in React Native. Can we avoid this somehow? Or just rely on firebase to trigger when the data changes?

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+    return Math.floor(er.timestamp / (range / CHART_POINTS_NUMBER))
+  }, [])
+
+  const localCurrencyCode = useLocalCurrencyCode()
+  const displayLocalCurrency = useCallback(
+    (amount: BigNumber.Value) =>
+      getLocalCurrencyDisplayValue(amount, localCurrencyCode || LocalCurrencyCode.USD, true),
+    [localCurrencyCode]
+  )
+  const goldToDollars = useCallback(useGoldToDollarAmount, [])
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: why wrap these in useCallback?

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+              data: chartData.map((el) => el.amount),
+            },
+          ],
+        }}
+        width={CHART_WIDTH}
+        height={CHART_HEIGHT}
+        chartConfig={{
+          backgroundGradientFrom: colors.background,
+          backgroundGradientTo: colors.background,
+          color: (opacity = 1) => `rgba(0, 0, 0)`,
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: why lambdas here and below?

File: packages/mobile/src/exchange/CeloGoldOverview.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+export function CeloGoldOverview({ t, testID }: Props) {
+  useBalanceAutoRefresh()
+  const localCurrencyCode = useLocalCurrencyCode()
+  const displayLocalCurrency = useCallback(
+    (amount: BigNumber.Value) =>
+      getLocalCurrencyDisplayValue(amount, localCurrencyCode || LocalCurrencyCode.USD, true),
+    [localCurrencyCode]
+  )
+
+  const goldToDollars = useCallback(useGoldToDollarAmount, [])
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldOverview.tsx]: ditto as above, why wrap these?

File: packages/mobile/src/exchange/CeloGoldOverview.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+  return (
+    <View style={styles.container} testID={testID}>
+      <Text style={styles.title}>{t('celoGold')}</Text>
+      <Text style={styles.balance}>
+        {goldBalance && (
+          <Text style={fontStyles.semiBold}>{getMoneyDisplayValue(goldBalance)}</Text>
+        )}
+      </Text>
+      <Text style={[fontStyles.light, styles.localBalance]}>
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldOverview.tsx]: Small nit, good to avoid unnecessary style arrays

File: packages/mobile/src/exchange/actions.ts [hunk 12]
```diff
@@ -1,13 +1,17 @@
...
 
 export enum Actions {
   FETCH_EXCHANGE_RATE = 'EXCHANGE/FETCH_EXCHANGE_RATE',
   SET_EXCHANGE_RATE = 'EXCHANGE/SET_EXCHANGE_RATE',
+  SYNC_CELO_GOLD_EXCHANGE_RATE_HISTORY = 'EXCHANGE/SYNC_CELO_GOLD_EXCHANGE_RATE_HISTORY',
+  UPDATE_CELO_GOLD_EXCHANGE_RATE_HISTORY = 'EXCHANGE/UPDATE_CELO_GOLD_EXCHANGE_RATE_HISTORY',
   EXCHANGE_TOKENS = 'EXCHANGE/EXCHANGE_TOKENS',
 }
 
+//
```
Comment (jmrossy) [inline packages/mobile/src/exchange/actions.ts]: Why all the empty comments in this file now?

File: packages/mobile/src/exchange/hooks.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+export function useExchangeRate() {
+  return getRateForMakerToken(
+    useSelector(getExchangeRatePair),
+    CURRENCY_ENUM.DOLLAR,
+    CURRENCY_ENUM.GOLD
+  )
+}
+
+export function useGoldToDollarAmount(amount: number | string | BigNumber | null) {
```
Comment (jmrossy) [inline packages/mobile/src/exchange/hooks.ts]: Can we use `BigNumber.Value` type here to be more concise?

File: packages/mobile/src/exchange/reducer.ts [hunk 14]
```diff
@@ -1,28 +1,89 @@
...
 export const exchangeRatePairSelector = (state: RootState) => state.exchange.exchangeRatePair
+export const exchangeHistorySelector = (state: RootState) => state.exchange.history
 
-export const reducer = (state: State | undefined = initialState, action: ActionTypes): State => {
+export const historyReducer = (
+  state: State['history'] | undefined = initialState.history,
+  action: ActionTypes
+): State['history'] => {
   switch (action.type) {
+    case Actions.UPDATE_CELO_GOLD_EXCHANGE_RATE_HISTORY:
```
Comment (jmrossy) [inline packages/mobile/src/exchange/reducer.ts:42]: why the switch cases here when this reducer is only used for the `UPDATE_CELO_GOLD_EXCHANGE_RATE_HISTORY` case below?

File: packages/mobile/src/utils/formatting.ts [hunk 15]
```diff
@@ -20,6 +21,19 @@ export const getMoneyDisplayValue = (
   return includeSymbol ? symbol + formattedValue : formattedValue
 }
 
+export const getLocalCurrencyDisplayValue = (
```
Comment (jmrossy) [inline packages/mobile/src/utils/formatting.ts:24]: How does this differ than `getMoneyDisplayValue` above? Can the two be merged?

File: packages/mobile/src/firebase/saga.ts [hunk 16]
```diff
@@ -217,6 +220,42 @@ export function* watchWritePaymentRequest() {
...
+  const latestExchangeRate = history.celoGoldExchangeRates[history.celoGoldExchangeRates.length - 1]
+
+  const exchangeRatesPromise = new Promise((resolve, reject) => {
+    firebase
+      .database()
+      .ref(`${EXCHANGE_RATES}/cGLD/cUSD`)
+      .orderByChild('timestamp')
+      // timestamp + 1 cause .startAt is inclusive
+      .startAt(latestExchangeRate ? latestExchangeRate.timestamp + 1 : 0)
+      .once('value', (snapshot) => {
```
Comment (jmrossy) [inline packages/mobile/src/firebase/saga.ts]: Why use once instead of allowing firebase to keep in sync?

**Review by jmrossy — COMMENTED** [review]
Overall looks good! See comments below  :)

File: packages/mobile/src/utils/formatting.ts (hunk 15 above)
Comment (i1skn) [reply packages/mobile/src/utils/formatting.ts:24]: `getMoneyDisplayValue` uses `currency`(cUSD, cGOLD), `getLocalCurrencyDisplayValue` uses `localCurrency`(EUR, etc.). It looks weird for me as well, but our current currency logic is separated between these two, so I decided to keep it this way.

File: packages/mobile/src/exchange/reducer.ts (hunk 14 above)
Comment (i1skn) [reply packages/mobile/src/exchange/reducer.ts:42]: This is a good catch!

File: packages/mobile/src/exchange/actions.ts (hunk 12 above)
Comment (i1skn) [reply packages/mobile/src/exchange/actions.ts]: That was my attempt to separate the actions, cause it's hard to follow which one is which. But that is more like a proposal, if you guys do not like it - happy to remove it :)

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx (hunk 7 above)
Comment (i1skn) [reply packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: I would take a look on how to avoid it using not `once` as you mentioned below! Thanks!

File: packages/mobile/src/utils/formatting.ts (hunk 15 above)
Comment (jeanregisser) [reply packages/mobile/src/utils/formatting.ts:24]: Yes we'll probably merge them once we tackle the local currency refactor.

**Review by jeanregisser — COMMENTED** [review]
Very nice work!! 👍 🎉 

As talked together, my only concern is with the direct access to the raw exchange rate data which will probably be a problem once we need to support different time ranges. Or if we increase the frequency of the rates data (currently set to 30 mins).

File: packages/mobile/src/exchange/actions.ts (hunk 12 above)
Comment (annakaz) [reply packages/mobile/src/exchange/actions.ts]: I would support removing them- was also confused by the empty comments

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+import { exchangeHistorySelector } from 'src/exchange/reducer'
+import { Namespaces, withTranslation } from 'src/i18n'
+import { LocalCurrencyCode } from 'src/localCurrency/consts'
+import { convertDollarsToLocalAmount } from 'src/localCurrency/convert'
+import { useLocalCurrencyCode } from 'src/localCurrency/hooks'
+import { getLocalCurrencyExchangeRate } from 'src/localCurrency/selectors'
+import useSelector from 'src/redux/useSelector'
+import { goldToDollarAmount } from 'src/utils/currencyExchange'
+import { getLocalCurrencyDisplayValue } from 'src/utils/formatting'
+// @ts-ignore
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx:20]: Why the ts-ignore?

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+    return Math.floor(er.timestamp / (range / CHART_POINTS_NUMBER))
+  }, [])
+
+  const localCurrencyCode = useLocalCurrencyCode()
+  const displayLocalCurrency = useCallback(
+    (amount: BigNumber.Value) =>
+      getLocalCurrencyDisplayValue(amount, localCurrencyCode || LocalCurrencyCode.USD, true),
+    [localCurrencyCode]
+  )
+  // const blackColor = useCallback(() => `rgba(0, 0, 0)`, [])
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: Nit: remove

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+  if (max - min < CHART_MIN_VERTICAL_RANGE) {
+    const offset = CHART_MIN_VERTICAL_RANGE - (max - min) / 2
+    domain = { y: [min - offset, max + offset] }
+  }
+
+  return (
+    <View style={styles.container} testID={testID}>
+      <View style={styles.goldPrice}>
+        <View>
+          <Text style={styles.goldPriceTitle}>Gold Price</Text>
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: we should i18n this

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+            {displayLocalCurrency(currentGoldRateInLocalCurrency)}
+          </Text>
+          <Text style={styles.goldPriceChange}>
+            {rateChange.gt(0) ? '▴' : '▾'} {rateChange.toFixed(2)} (
+            {rateChangeInPercentage.toFixed(2)}%)
+          </Text>
+        </View>
+      </View>
+      <VictoryGroup
+        domainPadding={{ y: [30, 30], x: [5, 5] }}
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx]: Nit: if this is constant maybe define it outside the render function for efficiency?

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+        singleQuadrantDomainPadding={false}
+        padding={{ left: 0 }}
+        width={CHART_WIDTH}
+        height={CHART_HEIGHT}
+        data={chartData.map((el) => el.amount)}
+        domain={domain}
+      >
+        <VictoryLine interpolation="monotoneX" />
+        {/* 
+        // @ts-ignore */}
```
Comment (jmrossy) [inline packages/mobile/src/exchange/CeloGoldHistoryChart.tsx:243]: why the ts ignore?

File: packages/mobile/src/exchange/reducer.ts (hunk 14 above)
Comment (jmrossy) [reply packages/mobile/src/exchange/reducer.ts:42]: do you still need to fix this?

**Review by jmrossy — COMMENTED** [review]
Almost ready, just a few  more small things!

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx (hunk 21 above)
Comment (i1skn) [reply packages/mobile/src/exchange/CeloGoldHistoryChart.tsx:243]: Cause victory charts library does not have type binding and next line is causing a type error.

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx (hunk 17 above)
Comment (i1skn) [reply packages/mobile/src/exchange/CeloGoldHistoryChart.tsx:20]: Cause victory charts library does not have type binding and next line causing a type error.

File: packages/mobile/src/exchange/reducer.ts (hunk 14 above)
Comment (i1skn) [reply packages/mobile/src/exchange/reducer.ts:42]: Not anymore, cause `UPDATE_CELO_GOLD_EXCHANGE_RATE_HISTORY` is the only event now related to `history`

**Review by jmrossy — APPROVED** [review]

File: packages/mobile/src/exchange/CeloGoldHistoryChart.tsx (hunk 17 above)
Comment (cmcewen) [reply packages/mobile/src/exchange/CeloGoldHistoryChart.tsx:20]: https://github.com/FormidableLabs/victory-native/issues/269
